### DISCUSSION TITLE
How to get something equivalent to a partially autodisposable double family provider?

### DISCUSSION THREAD

**Discussion by manuel-plavsic** [rrousselGit/riverpod#1493] [Q&A]
What I am trying to achieve:

A "double-family" provider with:
- autodisposable account identifier
- non-autodisposable some_key

```dart
// not possible
ref.read(provider(accountId, some_key);

// or:
// ref.read(provider(accountId)(some_key));
// (probably possible with Riverpod with nested providers, although it would mean having to write the type of the nested provider which is pretty long)
```

More top-level I would then be listening to the first component (`accountId`):
```dart
ref.listen(
  provider(widget.currentlySelectedAccount.id),
  (previous, next) {},
);
```

## Expectation

if I change account, I expect the provider to lose the information of the cached things.

## Incomplete solutions

Two possible, incomplete solutions (maybe wrong):
- Nested providers, as hinted above (even though I did not test it). However, I imagine that if Riverpod does not provider a simpler way to automate this, one probably has to write very long provider types. Also, I am not sure if this would be an anti-pattern.
- Another possible solution is to wrap the `accountId` and the `key` in a tuple or a custom object, however my desired goal will then not be achieved. In addition, such a provider needs to be either:
  - fully autodisposable (not an option, because everything will be garbage collected pretty quickly)
  - fully non-autodisposable, to ensure caching works. However, this means keeping the state of accounts I switched from, which is undesirable.
This case is maybe fixable if the provider is manually refreshed or invalidated when changing account. **Edit**: this fix is actually not feasable, because one cannot enter only `ref.read(provider(accountId))`, since also all "secondary" family keys would be needed to invalidate all providers.

## Question

What do you think would be a clean solution to this problem?

**Comment by manuel-plavsic** [comment]
I think the question may sound a bit too abstract, so I'll share my specific case.

`some_key` is a freezed data object containing two `int` fields month and year.

If I change the month and year pair from the calendar view, I download a list of events for a given month.

Therefore, the value held by `provider` (the provider name that was mentioned above) is of type `List<Event>`.

When I switch account, I would like to autodispose the month events tied to the previous account. I plan to do so by making sure the following happens.

```dart
// newCurrentlySelectedAccount's id differs from the prev. selected account's one
ref.listen(
  provider(widget.newCurrentlySelectedAccount.id),
  (previous, next) {},
);
// ... therefore prevCurrentlySelectedAccount no longer listened to.
```

**Comment by rrousselGit** [comment]
If you want to pass multiple values to a family provider, you can make a class (or tuple) that holds both. The important part is that this class implements ==

**Reply by manuel-plavsic** [reply]
Yes, I already did this. However, you cannot have a partially autodisposable tuple or class.

**Reply by rrousselGit** [reply]
I think I understand what you want.

Maybe you want to use two providers instead of one? You don't ave to use just one provider

**Reply by manuel-plavsic** [reply]
Would you use nested providers?

For example:

- an autodisposable family provider, e.g. called `topLevelProvider`, that takes account ids as arguments and holds state of a list/set of `nestedProvider`.
- a non-autodisposable family provider, e..g called `nestedProvider`, that takes monthAndYearPairs as arguments and holds state of `List<Event>`.

Or do you think that nested providers are an anti-pattern in Riverpod?

Would you rather encourage having two top level providers? I don't really see how I would accomplish this, though.

**Reply by rrousselGit** [reply]
No, that's an anti pattern.
Providers should only be top level.

What would be the issue with having two top level providers?

**Reply by manuel-plavsic** [reply]
I did not have an idea on how to achieve it using two globally defined providers. However, I think I might have figured it out (even though it still doesn't work).

Recall the agenda example...

- autodisposable, family `accountIdProvider` (family argument: `accountId`)
```dart
final accountIdProvider =
    Provider.autoDispose.family<void, int>((ref, id) {});
```
- autodisposable, family `monthEventsProvider` (family argument: class/tuple `accountId` and `monthAndYearPair`)
```dart
final monthEventsProvider = StateProvider.family
    .autoDispose<List<Event>, IdMonthAndYear>((ref, idMonthAndYear) {
  ref.read(accountIdProvider(idMonthAndYear.id).addListener(monthEventsProvider(idMonthAndYear), (previous, next) { }));
  ...
  // use idMonthAndYear.monthAndYear here...
  return someList;
});

// where

@freezed
class IdMonthAndYear with _$IdMonthAndYear {
  const factory IdMonthAndYear(
      {required int id,
      required MonthAndYear monthAndYear}) = _IdMonthAndYear;

  factory IdMonthAndYear.fromJson(Map<String, dynamic> json) =>
      _$IdMonthAndYearFromJson(json);
}
```

The `ref.read(accountIdProvider(idMonthAndYear.id).addListener(monthEventsProvider(idMonthAndYear), (previous, next) { }));` part was my guess, but it is not correct. I don't know how to programmatically make `accountIdProvider` listen to `monthEventsProvider`.Should I use `dependencies`, or what is the most pragmatic way?  Can you help me here?

**Reply by rrousselGit** [reply]
You probably want to use `ref.watch` or `ref.listen`

**Reply by manuel-plavsic** [reply]
Should I watch `monthEventsProvider` from `accountIdProvider`, or `accountIdProvider` from `monthEventsProvider`?

Watching/listenning `accountIdProvider` from `monthEventsProvider` does not work.

I am not so sure how to watch/listen  `monthEventsProvider` from `accountIdProvider`.

**Reply by manuel-plavsic** [reply]
This is why I had tried with `  ref.read(accountIdProvider(idMonthAndYear.id).addListener(monthEventsProvider(idMonthAndYear), (previous, next) { }));` thinking that addListener would use `ref.listen`

**Comment by manuel-plavsic** [comment]
I think another possibility would be to use an autodisposable statenotifierprovider, whose family type is `AccountId`, and the state is a map. But then I have create statenotifier methods just to manage the map (a bit what the family provider was designed for).

This is workaround probably solves my issue best (because it gets autodisposed properly). However, it forces me to handle maps myself.
